feat(event-handler): add resolution logic to base router

svozza opened the pull request:
## Summary
This PR implements the final resolution logic for the `BaseRouter` class. The `resolve` method is no longer abstract and has a concrete implementation where it looks up the route registry using the path and tries to retirve a resolver for the route. If not route is found, we throw a `NotFoundErroe`. If any exceptions are thrown from the handler, these are caught and handled by the already existing `handleError` method.

### Changes
- Update `BaseRouter` to have. oncrete implementation of `resolve` method
- Added new conversion functions to produce the `APIGatewayProxyResult` type that lambda expects
- Updated the unit tests now that the `resolve` function is implemented in the base class
- Added tests to ensure `this` context is preserved when using decorators.

**Issue number:** closes #4142 

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

svozza commented on packages/event-handler/src/rest/converters.ts:
```diff
@@ -45,4 +49,45 @@ export const proxyEventToWebRequest = (event: APIGatewayProxyEvent) => {
+    headers,
+    multiValueHeaders,
+    body: await response.text(),
+    isBase64Encoded: false,
```
I've hardcoded this for now but we will need to figure out how to deal with this

svozza commented on packages/event-handler/src/rest/BaseRouter.ts:
```diff
@@ -133,11 +139,49 @@ abstract class BaseRouter {
   }
 
-  public abstract resolve(
+  public async resolve(
```
Now that this function has an implementation, there are no abstract methos remaining. Should `BaseRouter` still be an abstract class?

dreamorosi replied:
If there's no longer a reason for it to be abstract, I'm ok to do away with it.

svozza commented on packages/event-handler/src/rest/BaseRouter.ts:
```diff
@@ -133,11 +139,49 @@ abstract class BaseRouter {
+
+    try {
+      const request = proxyEventToWebRequest(event);
+      const path = new URL(request.url).pathname as Path;
```
In general, I dislike casting, should these two assignments  use type guards?

dreamorosi replied:
I dislike it too.

My rule of thumb is that if this is a TypeScript type system limitation (aka TS can't detect the type even though it should) or we're reasonably sure that this type is the only realistic code path, then type casting can be ok.

If instead there's a runtime scenario in which the value might not be of that type, then a type guard is preferred.

dreamorosi replied:
For the line below, `as HttpMethod`, should also be fine since it comes directly from the request we just parsed.

svozza replied:
I've created a type guard for the HTTP method but I've moved the check outside the `try` block because if we throw a `MethodNotAllowedError`, inside the `catch` block, the error handling logic will try to create the `Request` object and throw with the `TypeError` mentioned above. Rather than throw a `MethodNotAllowedError`, we just return a `APIGatewayProxyResult`  with the appropriate 405 error code.

The creation of the Wed API `Request` object has also been moved out of the `catch` block. If this fails it is an unrecoverable error so we should just let the lambda fail.

dreamorosi reviewed (changes requested):
I know we added this in a previous PR, but after testing the `BaseRouter` in a Lambda function behind API Gateway REST I think I found a bug in the `proxyEventToWebRequest` function.

Specifically, the headers parsing (below), is problematic as it concatenates values if the same header is present in both `headers` and `multiValueHeaders` - which seems to be always the case in my tests:

```ts
for (const [name, value] of Object.entries(event.headers ?? {})) {
  if (value != null) headers.append(name, value);
}

for (const [name, values] of Object.entries(event.multiValueHeaders ?? {})) {
  for (const value of values ?? []) {
    headers.append(name, value);
  }
}
```

For example this request payload (shortened for brevity):

```json
{
  "resource": "/",
  "path": "/",
  "httpMethod": "GET",
  "headers": {
    "Host": "abcd1234.execute-api.eu-west-1.amazonaws.com",
    "X-Forwarded-Proto": "https"
  },
  "multiValueHeaders": {
    "Host": ["abcd1234.execute-api.eu-west-1.amazonaws.com"],
    "X-Forwarded-Proto": ["https"]
  },
  "queryStringParameters": null,
  "multiValueQueryStringParameters": null,
  "pathParameters": null,
  "stageVariables": null,
  "requestContext": { ... },
    "domainName": "abcd1234.execute-api.eu-west-1.amazonaws.com",
    "deploymentId": "abcd123",
    "apiId": "abcd1234"
  },
  "body": null,
  "isBase64Encoded": false
}
```

results in this:

```ts
console.log(headers.get('Host'));
// abcd1234.execute-api.eu-west-1.amazonaws.com, abcd1234.execute-api.eu-west-1.amazonaws.com 
console.log(headers.get('X-Forwarded-Proto'));
// https, https
```

making the URL parsing at L40 fail together with the request.

Additionally, like in this case, the `proxyEventToWebRequest` function is called and run twice both when first parsing the request and then again in the `except` block when trying to run the `handleError` - is there any way that we can run it only once?

Another thing I noticed is that when using API Gateway REST, the `requestContext.path` field includes the stage as path prefix, for example in a stage `prod` making a request to `stuff` would result in `"path":"/prod/stuff"`.

Do we expect customers to have to set it in as prefix at the router level, or in their route handlers? Should we instead just use the top level `path` and `resource` fields which in this case would be respectively `/stuff` and `/`?

svozza commented:
> I know we added this in a previous PR, but after testing the `BaseRouter` in a Lambda function behind API Gateway REST I think I found a bug in the `proxyEventToWebRequest` function.
> 
> Specifically, the headers parsing (below), is problematic as it concatenates values if the same header is present in both `headers` and `multiValueHeaders` - which seems to be always the case in my tests:

I've added logic to check whether the header already includes a value before we append it.

> Additionally, like in this case, the `proxyEventToWebRequest` function is called and run twice both when first parsing the request and then again in the `except` block when trying to run the `handleError` - is there any way that we can run it only once?
> 

I've moved the creation of the `Request` object out of the try/catch handler so we now only do it once.

> Another thing I noticed is that when using API Gateway REST, the `requestContext.path` field includes the stage as path prefix, for example in a stage `prod` making a request to `stuff` would result in `"path":"/prod/stuff"`.
> 
> Do we expect customers to have to set it in as prefix at the router level, or in their route handlers? Should we instead just use the top level `path` and `resource` fields which in this case would be respectively `/stuff` and `/`?

Agreed, I've changed it to use the top level  fields instead.
